### ISSUE TITLE
[Extension]: r0mar0ma/sd-webui-pez-dispenser can't spawn a tab. dev but not new 

### ISSUE THREAD

**Issue by tau0-deltav** [vladmandic/automatic#2742]
### Issue Description

Hello, I've not actually used github except completely passively before so forgive me if this isn't quite right.
[This extension](https://github.com/r0mar0ma/sd-webui-pez-dispenser) is tragically overlooked (doesn't show up on a1111's extension list) tool that implements Hard Prompts Made Easy. This is an inherently Just and Good thing that replaces 'prompt engineers' with AdamW. God knows I was getting sick of GregR.

**What is supposed to happen that does not happen?** A tab that lets you run the prompt optimiser independently of the image pipeline - this is the bit which I'd hoped would never break for that exact reason. All the extensions methods can still (technically) be called from the script view. **The inconvenience of doing so is beyond the scope of this issue**.

The extension has some faults it inherits from A1111. I might look at evangelising for upstream/hacking together fixes for those myself. **This bug is one that was actually predicted by the changelog for SD.Next and I'm reporting it because I was asked most politely to:** 
> 2023-01-23
> major internal ui module refactoring
this may cause compatibility issues if an extension is doing a direct import from ui.py
in which case, report it so we can add a compatibility layer

Yup, here you go, when the extension is set to (using the only option it adds to the settings menu) "tab and script" or "tab only" mode, the installer does something unorthodox to get a (seemingly unremarkable) tab slot, which leaves this in the console:
```
20:34:58-450177 INFO     Load UI theme: name="black-teal" style=Auto base=sdnext.css                          
20:34:59-402344 ERROR    executing callback:                                                                  
                         /home/nodox/automatic/extensions/sd-webui-pez-dispenser/scripts/pez-dispenser.py
                         ui_tabs_callback: AttributeError                                                     
╭──────────────────────────────────── Traceback (most recent call last) ─────────────────────────────────────╮
│ /home/nodox/automatic/modules/script_callbacks.py:207 in ui_tabs_callback                           
│                                                                                                            │
│   206 │   │   │   t0 = time.time()                                                                         │
│ ❱ 207 │   │   │   res += c.callback() or []                                                                │
│   208 │   │   │   timer(t0, c.script, 'ui_tabs')                                                           │
│                                                                                                            │
│ /home/nodox/automatic/extensions/sd-webui-pez-dispenser/scripts/pez-dispenser.py:433 in create_tab
│                                                                                                            │
│   432 │   │   process_image_button.click(                                                                  │
│ ❱ 433 │   │   │   ui.wrap_queued_call(inference_image),                                                    │
│   434 │   │   │   _js = "pezdispenser_submit",                                                             │
╰────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
AttributeError: module 'modules.ui' has no attribute 'wrap_queued_call'
20:34:59-860635 INFO     Local URL: http://127.0.0.1:7861/
```
Please don't hesitate to ask for any more info and I can probably use my decade-atrophied programming abilities to do some more research, getting in to this generative AI stuff is slowly teaching me computing in general again :) 
Maybe one day I'll manage a pull request. For now, thanks for your (incredibly) hard work. Hope this is a trivial fix.

Just while you're reading - (should make this a Discussion?)
Any opinions on the suitability of this extension, [reference code](https://github.com/YuxinWenRick/hard-prompts-made-easy) or just the [paper's core idea](https://arxiv.org/abs/2302.03668) for integration in to SD.Next? It's good yes? Just me? Not on anyone else's back of course - just wondering if it would 'fit' if I were to do it. `torch.compile()` is not an outlandish possibility here to my eyes but... maybe I've got bad eyes.

### Version Platform Description

 (i've been seeing this since I decided to switch - this replicates in unmodified SD.next, not sure about the branch.)
(triton/xformers/torchao/stable-fast from source may well cause all my other issues but not this.)
```
automatic]$ ./webui.sh --debug --experimental --use-cuda --use-xformers 
Create and activate python venv (happens whether or not I `source` first
Launching launch.py...
20:39:44-804066 INFO     Starting SD.Next                                                                                                                                               
20:39:44-805964 INFO     Logger: file="/home/nodox/automatic/sdnext.log" level=DEBUG size=64 mode=create                                                                                  
20:39:44-806681 INFO     Python 3.11.6 on Linux                                                                                                                                         
20:39:44-816070 INFO     Version: app=sd.next updated=2024-01-23 hash=edf1dc68 url=https://github.com/vladmandic/automatic//tree/dev                                                    
20:39:45-758317 INFO     Latest published version: 5fb290f443d5f38a5c9f6e6095aabeab8e3a991d 2024-01-13T13:47:29Z                                                                        
20:39:45-760067 INFO     Platform: arch=x86_64 cpu= system=Linux release=6.7.0-zen3-1-zen python=3.11.6                      

```
Firefox may well be responsible for the squashed script mode ui (which I'd like to see fixed but am not reporting here) but **given that this bug happens before the browser even tries to connect** - we can rule it out as a cause here.
The script view (if someone else wouldn't mind taking a look at it for me - just in text2img where you'd otherwise pick XYZ grid and such) being so mangled does exacerbate this issue - I'd been able to accept relying on the tab view instead. Not asking for a fix, just a view on where the 'problem' is - including possibly my browser.

### URL link of the extension

https://github.com/r0mar0ma/sd-webui-pez-dispenser

### URL link of the issue reported in the extension repository

This issue is newer than the last update to that repo and was expected on this end, so not done..

### Acknowledgements

- [X] I have read the above and searched for existing issues

**Comment by vladmandic** [comment]
i've added modules.ui.wrap_queued_call compatibility method, if there is anything else let me know (i cant test the extension end-to-end right now).

regarding making it a built-in functionality, i'm afraid that sheer complexity of this extension doesn't make it a candidate.

btw, you're saying it doesn't appear in extension list - @Aptronymist can you add it to your list so it'll merge?

**Comment by tau0-deltav** [comment]
You're killing it vlad. Fix confirmed, all previous functionality is present.
Yesterday I also noticed the "TE2:" text encoder selection in the changelog as well. This makes this extension more powerful for SDXL here: we can generate prompts for the ViT-L and ViT-bigG encoders separately. Meaningful especially given one is ~32 times faster (per initialisation) than the other. Will experiment.

Thank you!

**Comment by Aptronymist** [comment]
> i've added `modules.ui.wrap_queued_call` compatibility method, if there is anything else let me know (i cant test the extension end-to-end right now).
> 
> regarding making it a built-in functionality, i'm afraid that sheer complexity of this extension doesn't make it a candidate.
> 
> btw, you're saying it doesn't appear in extension list - @Aptronymist can you add it to your list so it'll merge?

weird, how did I not see this earlier.

**Comment by StableInfo** [comment]
Hello to all:
1) I had first mentioned it here (https://github.com/vladmandic/automatic/issues/2742) Do you remember @vladmandic , it seems this was important and still is.
2) What is "prompt engineers' with AdamW" @tau0-deltav ? I googled it I could not find it. I am afraid I am missing on something important that I don't know about?
3) Same question with "GregR" You mentioned it but I am afraid I am missing on somemthing else.
4) Do you explaining to me what was said on this post? If I am correct a fix was made by Vlad, but do I have to code something or is it already available for standard installs? Second thing, could you EPXLAIN MORE what you meant by "TE2:" ? I don't know nothing about it. But I think you mentioned ViT-L  which I think I recognize from the PEZ extension its one of the options when generating prompts if I am not wrong?

Please take 3-5 minutes of your time to inform me/update me so I can get in the bandwagon. I had already knew about pez but Paused for a while so I might be outdated in my infos.

**Comment by Aptronymist** [comment]
> 4\. Second thing, could you EPXLAIN MORE what you meant by "TE2:" ? I don't know nothing about it.

"TE2:" is what you can type into the prompt to specifically put things into the secondary text encoder (TE2) for SDXL, anything after "TE2:" goes to that encoder specifically. AFIAK your single prompt goes to both text encoders by default.

**Comment by tau0-deltav** [comment]
Not related to this reply: Slightly embarrassed to be replying to anything on this github given my recent temper tantrum. I need to stop saying up past my bedtime.

_Nearly done writing this and it's just struck me: If you have questions like this you could try asking an LLM? They understand english better than I do.

@StableInfo. First of all you could have googled "prompt engineer" and "AdamW". But:

[Hard Prompts Made Easy: Gradient-Based Discrete Optimization for Prompt Tuning and Discovery](https://doi.org/10.48550/arXiv.2302.03668)

Here's the paper I linked in the first post. This is fairly detailed scientific writing by the people who designed the (base codebase of) this plug-in. This answers all of the actually important questions. I think.  About 60% of it went over my head. **This is what prompt optimis(z)ation is.** for this context.

- **On the other hand**, I wrote much of this with my [tongue in my cheek](https://en.wiktionary.org/wiki/tongue-in-cheek#Translations) and my favourite contaminants in circulation. 
- Google is not a good place to learn any more. Unless you broke your w key and can't find another way to Wikipedia.

> What is "prompt engineers' with **AdamW**" @tau0-deltav ? I googled it I could not find it. I am afraid I am missing on something important that I don't know about?
### AdamW
I'm not sure how you managed to turn up nothing for AdamW. [Wikipedia has a section for Adam](https://en.wikipedia.org/wiki/Stochastic_gradient_descent#Adam) under Stochastic Gradient Descent. [With this reference to AdamW](https://arxiv.org/abs/1711.05101). It's a very popular algorithm.

AdamW is an improvement of Adam. Adam is SGD except more complex, memory-intensive and _" subsequent analysis has revealed that Adam does not converge for all convex objectives"_. 
But it's faster for lots of real world things.

 In this particular case the AdamW in question is the [pytorch reference](https://pytorch.org/docs/stable/optim.html#base-class) one here: [torch.optim.AdamW](https://pytorch.org/docs/stable/generated/torch.optim.AdamW.html).  
 
### Prompt Engineers, Prompt Engineering, Gregg Rutkowski

> What is "**prompt engineers**' with "AdamW"
> Same question with "GregR" You mentioned it but I am afraid I am missing on something else.

<details><summary>This is a bad joke</summary>
GregR is truly my very own Turboencabulator. Greg R refers to [Grzegorz Rutkowski](https://www.artofmtg.com/artist/grzegorz-rutkowski/) 
A prompt engineer is a person who does prompt engineering. Supposedly. I do not take these people seriously and I think they are probably some sort of multi-level-marketing/pyramid scheme thing. Maybe crypto-bros who didn't sell the GPUs?
Their 'job' is to 'program' generative AIs by getting the right vibes from the text encoder. It's like... horse whispering except no animals are soothed so strictly worse.
I'm not going to explain the Greg Rutkowski meme to you. 
I was claiming that This add-on replaced these people because it was better than putting "Greg Rutkowski" in front of your prompt. Which it is. Sometimes it might actually decide on Greg Rutkowski anyway, in principle.
</details> 
Thus AdamW > GregR. You see? its not funny. :(

**Comment by vladmandic** [comment]
FYI and IMO, main problem with sdxl-dual clip is that results are concatenated.
TE1 is simple old CLIP ViT-L and TE2 is OpenCLIP ViT-bigG
but since results of OpenCLIP get contatenated to results of CLIP-ViT, you have a problem that poorer clip accounts for majority of guidance.
negative embeds and all that stuff is a story for another day.

**Comment by Aptronymist** [comment]
> FYI and IMO, main problem with sdxl-dual clip is that results are concatenated. TE1 is simple old CLIP ViT-L and TE2 is OpenCLIP ViT-bigG but since results of OpenCLIP get contatenated to results of CLIP-ViT, you have a problem that poorer clip accounts for majority of guidance. negative embeds and all that stuff is a story for another day.

Then why send anything to CLIP ViT-L at all? If we just put TE2: at the start of the prompt, that should take care of it, no?

**Comment by vladmandic** [comment]
try :)

**Comment by Aptronymist** [comment]
> @Aptronymist Specifically, one of them speaks CLIP ViT-L and the other speaks OpenCLIP ViT-bigG. the latter being a MUCH MUCH _MUCH_ more capable text encoder/language model than CLIP ViT-L It's not just that it's seen more things it's that it just... can actually read. You speak to it in plain english when you're using SD normally and not like a psychopath like the PEZ DISPENSER lets you.
> 
> No idea which one of those TE2: _is_ though. Or which one is considered the secondary CFG scale. Or what weighted or pooled embeds do. Or how negative guidance is invoked.
> 

As far as I know, TE2 is the second, ViT-G model that is superior, I will try prompting with just that (`TE2: 1girl, astronomical boobs, etc, etc`) to see what happens.

> [Hard Prompts Made Easy: Gradient-Based Discrete Optimization for Prompt Tuning and Discovery](https://doi.org/10.48550/arXiv.2302.03668)
> 
> Here's the paper I linked in the first post. This is fairly detailed scientific writing by the people who designed the (base codebase of) this plug-in. This answers all of the actually important questions. I think. About 60% of it went over my head. **This is what prompt optimis(z)ation is.** for this context.

I actually tried the PEZ Dispenser extension out today myself a little, but I need to read the docs and all before I understand it of course. It was functional as far as I could tell, made images with psychotic prompts alright!
